src/transformation/visitors/call.ts

```diff
@@ -16,13 +16,75 @@ import { transformLuaTableCallExpression } from "./lua-table";
 
 export type PropertyCallExpression = ts.CallExpression & { expression: ts.PropertyAccessExpression };
 
+function getExpressionsBeforeAndAfterFirstSpread(
+    expressions: readonly ts.Expression[]
+): [readonly ts.Expression[], readonly ts.Expression[]] {
+    // [a, b, ...c, d, ...e] --> [a, b] and [...c, d, ...e]
+    const index = expressions.findIndex(ts.isSpreadElement);
+    const hasSpreadElement = index !== -1;
+    const before = hasSpreadElement ? expressions.slice(0, index) : expressions;
+    const after = hasSpreadElement ? expressions.slice(index, expressions.length) : [];
+    return [before, after];
+}
+
+function transformSpreadableExpressionsIntoArrayConcatArguments(
+    context: TransformationContext,
+    expressions: readonly ts.Expression[] | ts.NodeArray<ts.Expression>
+): lua.Expression[] {
+    // [...array, a, b, ...tuple()] --> [ [...array], [a, b], [...tuple()] ]
+    // chunk non-spread arguments together so they don't concat
+    const chunks: ts.Expression[][] = [];
+    for (const expression of expressions) {
+        if (ts.isSpreadElement(expression)) {
+            chunks.push([expression]);
+            const next = expressions[expressions.indexOf(expression) + 1];
+            if (next && !ts.isSpreadElement(next)) {
+                chunks.push([]);
+            }
+        } else {
+            const lastChunk = chunks[chunks.length - 1];
+            lastChunk.push(expression);
+        }
+    }
+
+    return chunks.map(chunk => wrapInTable(...chunk.map(expression => context.transformExpression(expression))));
+}
+
+export function flattenSpreadExpressions(
+    context: TransformationContext,
+    expressions: readonly ts.Expression[]
+): lua.Expression[] {
+    const [preSpreadExpressions, postSpreadExpressions] = getExpressionsBeforeAndAfterFirstSpread(expressions);
+    const transformedPreSpreadExpressions = preSpreadExpressions.map(a => context.transformExpression(a));
+
+    // Nothing special required
+    if (postSpreadExpressions.length === 0) {
+        return transformedPreSpreadExpressions;
+    }
+
+    // Only one spread element at the end? Will work as expected
+    const [firstExpression] = postSpreadExpressions;
+    if (postSpreadExpressions.length === 1 && ts.isSpreadElement(firstExpression)) {
```

Comment:
isn't `ts.isSpreadElement(firstExpression)` always true by definition of `getExpressionsBeforeAndAfterFirstSpread`?

Reply:
true, that can be refactored